### DISCUSSION TITLE
Computing effective elastic moduli

### DISCUSSION THREAD

**Discussion by clariu25** [idaholab/moose#23460] [Q&A Modules: Solid mechanics]
Hi,
I'm trying to retrieve the bulk and shear moduli from a simple cube (homogeneous). The idea is to later add porosity and evaluate its effective moduli, therefore, I can't extract the values from the elasticity tensor, as that remains constant. For now I'm only applying some load at the top (ZZ direction). I'm using the Young modulus and Poisson ratio as inputs for the `ComputeIsotropicElasticityTensor`.

By evaluating the stress/strain relations I should be able to retrieve the same values for the elastic moduli that I used as inputs, as the material hasn't changed, but that is not case. Non of the calculated moduli coincide with the input ones.
Results (Modulus - expected value - output value - calculation):
- K - 36GPa  - 30GPa     -   hydrostatic_stress/volumetric_stress
- G - 72MPa -144MPa   -  stress_xz/strain_xz (exactly twice as the input file value)
- E - 22MPa  - 18MPa   -  stress_zz/strain_zz
- Poisson - 0.499 - 0.01  - strain_yy/strain_zz

Has anyone encounter something similar? Or is my assumption of the non-changed material properties wrong?
Thank you in advance for your replies.

Regards,
Clara

**Comment by hugary1995** [comment]
Can you elaborate on how you retrieve the moduli "by evaluating the stress/strain relations"?

**Reply by clariu25** [reply]
The relations that allow to calculate the elastic moduli. The equations that I wrote next to the results.

**Reply by GiudGiud** [reply]
I saw you closed/re-opened this.

are you relying on the material property output to retrieve these values?
do you have the input snippets?

**Reply by clariu25** [reply]
Hello, Yes, I wanted to delete this an open a new discussion but I figured it makes more sense to continue here.
Yes, so I'm focusing on the bulk and the shear modulus and to obtain them I'm using Aux variables and aux kernels. 

```
[AuxVariables]
   [volum_stress]
      family = MONOMIAL
      order = SECOND
    []
    [stress_xz]
      family = MONOMIAL
      order = SECOND
    []
    [volum_strain]
      family = MONOMIAL
      order = SECOND
    []
    [strain_xz]
      family = MONOMIAL
      order = SECOND
    []
    [K]
      family = MONOMIAL
      order = SECOND
    []
    [G]
      family = MONOMIAL
      order = SECOND
    []
 []
 [AuxKernels]
   [vol_stress]
      type = RankTwoScalarAux
      rank_two_tensor = stress
      variable = volum_stress
      scalar_type = Hydrostatic
      execute_on = timestep_end
    []
    [stress_xz]
      type = RankTwoAux
      variable = stress_xz
      rank_two_tensor = stress
      index_i = 0
      index_j = 2
      execute_on = timestep_end
    []
    [vol_strain]
      type = RankTwoScalarAux
      rank_two_tensor = total_strain
      variable = volum_strain
      scalar_type = VolumetricStrain
      execute_on = timestep_end
    []
    [strain_xz]
      type = RankTwoAux
      variable = strain_xz
      rank_two_tensor = total_strain
      index_i = 0
      index_j = 2
      execute_on = timestep_end
    []
    [Bulk_m]
      type = ParsedAux
      variable = 'K'
      coupled_variables = 'volum_strain volum_stress'
      expression = 'abs(volum_stress/volum_strain)'
      execute_on = 'TIMESTEP_END'
    []
    [Shear_m]
      type = ParsedAux
      variable = 'G'
      coupled_variables = 'stress_xz strain_xz'
      expression = 'stress_xz/(2*strain_xz)'
      execute_on = 'TIMESTEP_END'
    []
[] 
```
So far the Shear modulus remains constant, but K has a very odd behavior, reaching values extremely high at the end of the temperature plume. In the picture I show a "ring" that appears at the end of the temperature plume, where K reaches values in the order of 10^4GPa. The stress and strain show similar behavior so I don't understand where does this "ring" comes from
(I multiply the volumetric strain by -1 for visualization purposes)
![image](https://github.com/user-attachments/assets/e8649530-c13f-4261-ae91-130e469aecef)

**Reply by GiudGiud** [reply]
> So far the Shear modulus remains constant, but K has a very odd behavior, r

so of the 4 quantities, is K the only one that's off now?

**Reply by clariu25** [reply]
Yes, but as I'm using the volumetric strain and stress to compute K, there might be something wrong with them as well

**Reply by GiudGiud** [reply]
ok I think it's time for @hugary1995 to come back to help here :)

**Reply by hugary1995** [reply]
Try changing `vol_strain` to
```python
    [vol_strain]
      type = RankTwoScalarAux
      rank_two_tensor = total_strain
      variable = volum_strain
      scalar_type = Hydrostatic
      execute_on = timestep_end
    []
```
`VolumetricStrain` is probably not what you want/thought it to be.

**Reply by hugary1995** [reply]
I was about to refer you to the documentation: https://mooseframework.inl.gov/source/utils/RankTwoScalarTools.html, but now I see some potentially confusing description there... We are also missing the description for `VolumetricStrain` which is why you made the mistake in the first place -- not your fault at all.

**Reply by GiudGiud** [reply]
@hugary1995 would you mind creating an issue? We should track this

**Reply by hugary1995** [reply]
sure, just did

**Reply by clariu25** [reply]
I tried @hugary1995 's suggestion but the ring is still appearing when computing K, and now with even higher values. At least the behavior of strain and stress follow a more similar trend. I took the volumetric strain before as that is the definition of Bulk modulus, but probably as you mentioned I got confused with the descriptions. 
<img width="736" alt="image" src="https://github.com/user-attachments/assets/c63573cf-3c6e-4a3d-8f4e-9350f5c7a90e" />

**Reply by hugary1995** [reply]
You said in the original post this is a homogeneous cube, so I don't see why you would see such a ring.

**Reply by clariu25** [reply]
Yes, this is an homogeneous cube and I'm injecting hot water in the center of the ring. I also do not expect this ring and I do not know what creates it. I think it is a numerical issue as when I refine the mesh the ring gets finer, but it is never completely gone. The graphs of strain and strain also look "fine", reasonable, they have similar behaviors (as they should), but I still get this ring when calculating the bulk modulus.

**Reply by clariu25** [reply]
I found the issue. I was dividing by zero (K=stress/strain): the deformation is too large, and it goes from negative to positive values. I get the ring at the point it crosses zero. 
I still have the issue regarding the wrong value of K previous to the injection

**Reply by hugary1995** [reply]
What's the purpose of this exercise? If you want to do homogenization, just do it the standard way.
